sklearn/metrics/_regression.py

```diff
@@ -791,25 +856,50 @@ def r2_score(y_true, y_pred, *, sample_weight=None, multioutput="uniform_average
     denominator = (
         weight * (y_true - np.average(y_true, axis=0, weights=sample_weight)) ** 2
     ).sum(axis=0, dtype=np.float64)
-    nonzero_denominator = denominator != 0
-    nonzero_numerator = numerator != 0
-    valid_score = nonzero_denominator & nonzero_numerator
-    output_scores = np.ones([y_true.shape[1]])
-    output_scores[valid_score] = 1 - (numerator[valid_score] / denominator[valid_score])
-    # arbitrary set to zero to avoid -inf scores, having a constant
-    # y_true is not interesting for scoring a regression anyway
-    output_scores[nonzero_numerator & ~nonzero_denominator] = 0.0
+
+    return _assemble_r2_explained_variance(
+        numerator=numerator,
+        denominator=denominator,
+        nb_outputs=y_true.shape[1],
+        multioutput=multioutput,
+        force_finite=force_finite,
+    )
+
+
+def _assemble_r2_explained_variance(
+    numerator, denominator, nb_outputs, multioutput, force_finite
+):
+    """Common part used by explained variance score and :math:`R^2` score"""
+
+    if not force_finite:
+        # Standard formula, that may lead to NaN or -Inf
+        output_scores = 1 - (numerator / denominator)
```

Comment:
Do we want to prevent raising a `RuntimeWarning` then? Since one has to force the parameter, it could expect to get non-finite values then.

Reply:
This is indeed a good idea, but there might be use cases where seeing the warning might still be useful. I would therefore suggest to let this to the wrapping caller for now.